### PR TITLE
Handle EBEROs: Use DelegationDate instead of ContractTerminated

### PR DESCRIPTION
**Pull request by aaomidi**
Solution for #1174

Public Suffix List (PSL) Pull Request (PR) Template
====

Each PSL PR needs to have a description, rationale, indication of DNS validation and syntax checking, as well as a number of acknowledgements from the submitter.  This template must be included with each PR, and the submitting party MUST provide responses to all of the elements in order to be considered.

### Checklist of required steps

* [x] Description of Organization
* [x] Robust Reason for PSL Inclusion
* [x] DNS verification via dig
* [x] Run Syntax Checker (make test)

* [x] Each domain listed in the PRIVATE section has and shall maintain at least two years remaining on registration, and we shall keep the \_PSL txt record in place in the respective zone(s) in the affected section

__Submitter affirms the following:__ 

  * [x] We are listing *any* third-party limits that we seek to work around in our rationale such as those between IOS 14.5+ and Facebook (see [Issue #1245](https://github.com/publicsuffix/list/issues/1245) as a well-documented example)

  * [x] This request was _not_ submitted with the objective of working around other third-party limits

  * [x] The [Guidelines](https://github.com/publicsuffix/list/wiki/Guidelines) were carefully _read_ and _understood_, and this request conforms
  * [x] The submission follows the [guidelines](https://github.com/publicsuffix/list/wiki/Format) on formatting and sorting

---

For Private section requests that are submitting entries for domains that match their organization website's primary domain, please understand that this can have impacts that may not match the desired outcome and take a long time to rollback, if at all.

To ensure that requested changes are entirely intentional, make sure that you read the affectation and propagation expectations, that you understand them, and confirm this understanding. 

PR Rollbacks have lower priority, and the volunteers are unable to control when or if browsers or other parties using the PSL will refresh or update.

(Link: [about propagation/expectations](https://github.com/publicsuffix/list/wiki/Guidelines#appropriate-expectations-on-derivative-propagation-use-or-inclusion))

 * [x] *Yes, I understand*.  I could break my organization's website cookies etc. and the rollback timing, etc is acceptable.  *Proceed*.
---

Description of Organization
====

Organization Website:  N/A

Reason for PSL Inclusion
====

Number of users this request is being made to serve: N/A

`.wed` is an EBERO TLD. This means it's still an active suffix for domains (e.g. may27.wed). I've updated the updater code to ensure that these DNS names (currently only one) don't get dropped by the automation until they're ready to be dropped.

DNS Verification via dig
=======

```
dig soa +short wed.
dns1.emdns.uk. hostmaster.nominet.org.uk. 2100001071 7200 900 2419200 3600
```

```
whois wed.
% IANA WHOIS server
% for more information on IANA, visit http://www.iana.org
% This query returned 1 object

domain:       WED

organisation: Emergency Back-End Registry Operator Program - ICANN
address:      12025 Waterfront Drive
address:      Suite 300
address:      Los Angeles CA 90094
address:      United States of America (the)

contact:      administrative
name:         EBERO Event Director
organisation: Internet Corporation for Assigned Names and Numbers (ICANN)
address:      12025 Waterfront Drive
address:      Suite 300
address:      Los Angeles CA 90094
address:      United States of America (the)
phone:        +1 310 301 5800
fax-no:       +1 310 823 8649
e-mail:       eventdirector@icann.org

contact:      technical
name:         Nominet EBERO Event Manager
organisation: Nominet
address:      Minerva House
address:      Edmund Halley Road
address:      Oxford Science Park
address:      Oxford Oxon OX4 4DQ
address:      United Kingdom of Great Britain and Northern Ireland (the)
phone:        +44 1865 332470
e-mail:       ebero@nominet.uk

nserver:      DNS1.EMDNS.UK 213.248.217.153 2a01:618:401:0:0:0:0:153
nserver:      DNS2.EMDNS.UK 103.49.81.153 2401:fd80:401:0:0:0:0:153
nserver:      DNS3.EMDNS.UK 213.248.221.153 2a01:618:405:0:0:0:0:153
nserver:      DNS4.EMDNS.UK 2401:fd80:405:0:0:0:0:153 43.230.49.153
nserver:      DNSA.EMDNS.UK 156.154.100.3 2001:502:ad09:0:0:0:0:3
nserver:      DNSB.EMDNS.UK 156.154.101.3 2001:502:2eda:0:0:0:0:3
nserver:      DNSC.EMDNS.UK 156.154.102.3 2610:a1:1009:0:0:0:0:3
nserver:      DNSD.EMDNS.UK 156.154.103.3 2610:a1:1010:0:0:0:0:3
ds-rdata:     60571 8 2 32ed00d841e724eaa980472283420ee9de48905e1b04709e79fc418c08fa1561

whois:        whois.nic.wed

status:       ACTIVE
remarks:      Registration information:
remarks:      https://www.icann.org/resources/pages/ebero-2013-04-02-en

created:      2013-12-12
changed:      2021-11-11
source:       IANA
```

Results of Syntax Checker (`make test`)
=========

```
PASS: libpsl_icu_load_dafsa_fuzzer
PASS: libpsl_icu_fuzzer
PASS: libpsl_icu_load_fuzzer
============================================================================
Testsuite summary for libpsl 0.21.2
============================================================================
# TOTAL: 3
# PASS:  3
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
Making check in tests
  CC       test-is-public-all.o
  CC       test-is-public.o
  CC       test-is-cookie-domain-acceptable.o
  CC       test-is-public-builtin.o
  CC       test-registrable-domain.o
  CCLD     test-is-cookie-domain-acceptable
  CCLD     test-is-public-builtin
  CCLD     test-is-public
  CCLD     test-is-public-all
  CCLD     test-registrable-domain
PASS: test-is-public-builtin
PASS: test-is-public
PASS: test-is-cookie-domain-acceptable
PASS: test-registrable-domain
PASS: test-is-public-all
============================================================================
Testsuite summary for libpsl 0.21.2
============================================================================
# TOTAL: 5
# PASS:  5
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
Making check in msvc

```

### CODE REVIEW

**Comment by dnsguru** [comment]
Thank you for stepping up to attempt assistance.

Needs some tweaks

This would lilely relist almost 100 .brands and PROBABLY needs to consider
BOTH vs either.  Names that have been withdrawn by the assignee will have
ContractTerminated and DelegationDate.

**Comment by aaomidi** [comment]
I did end up running the script and it only brought in .wed

I’m good with making a change that checks both the fields as AND. Or wouldn’t work as it would exclude TLDs like wed.

**Comment by aaomidi** [comment]
I've updated this PR with what was suggested. I think this might be good to go? I'll try to take another more holistic look into this script later.

**Comment by dnsguru** [comment]
> I've updated this PR with what was suggested. I think this might be good to go? I'll try to take another more holistic look into this script later.

@aaomidi thank you for this - typically I let @weppos review these.   Candidly, this is helpful but .wed is such an edge-case TLD as being the only EBERO listing - it might just be more efficient to add it into the legacy list with a comment and submit a PR to include it.  These really do not occur with any great frequency.

**Comment by aaomidi** [comment]
I kinda took a look at what zlint was doing. I think the original-previous to this PR assumption was technically incorrect. 

Honestly the blame here is with ICANN/IANA for not having an appropriate endpoint for this that doesn't have ambiguous endpoints.

I've been slightly tempted to also just check the root DNS zones for preparing this list lol.

**Comment by aaomidi** [comment]
Anymore thoughts on this? I still think based on the changes it's pulling through it would be a worthwhile addition.

**Comment by dnsguru** [comment]
Thanks Amir - volunteer's pace here vs corporate timelines (so this won't
happen quickly, to set expectations)

Code like this typically gets a few sets of eyes, including @weppos so that
we are not breaking anything in automation

**Comment by dnsguru** [comment]
This did not get attention due to resourcing, and now is behind a need to rebase if we wanted to clear the spider webs from it.  I'd like to thank people for the effort.  We also had some challenges with ICANN hitting the 10 year renewal mark on some of the nTLDs that we've mitigated with CPU's help.

suggest we close this

**Comment by dnsguru** [comment]
FYI .DESI is moving to EBERO https://www.icann.org/en/announcements/details/desi-to-be-placed-in-the-emergency-back-end-registry-operator-program-16-10-2023-en

**Comment by aaomidi** [comment]
Could we reopen this @dnsguru? I've ran the code again and it correctly put both `.desi` and `.wed` in the generated PSL.

I feel like we're causing unnecessary stress for folks using these unfortunate gTLDs by making a ton of automations break for them.

**Comment by Zenexer** [comment]
This is only going to become more of an issue as time goes on, so ideally this should be reopened.  The changes look good to me, although I'd add a comment somewhere explaining why the DelegationDate check is necessary, since it's not obvious from reading the code.

File: tools/newgtlds.go [hunk 1]
```diff
@@ -346,14 +349,15 @@ func getData(url string) ([]byte, error) {
...
+// ContractTerminated equal to true and an empty DelegationDate,
+// or a non-empty RemovalDate.
 func filterGTLDs(entries []*pslEntry) []*pslEntry {
 	var filtered []*pslEntry
 	for _, entry := range entries {
 		if _, isLegacy := legacyGTLDs[entry.ALabel]; isLegacy {
 			continue
 		}
-		if entry.ContractTerminated {
+		if entry.ContractTerminated && entry.DelegationDate == "" {
```
[the remark this answers is not in the document]
Comment (aaomidi) [reply tools/newgtlds.go:360]: Thank you! Changes made.

**Comment by aaomidi** [comment]
Sounds good @cpu, but can you reopen the PR? I think the diffs aren't updating since the PR is closed.

**Comment by cpu** [comment]
>  can you reopen the PR? 

I can't sorry, I don't have any special permissions here. @dnsguru, @weppos: Could one of you do that?

**Comment by dnsguru** [comment]
image
Won't let me re-open either - it is not a permissions thing

**Comment by cpu** [comment]
Weird! Thanks for checking. Probably best to open a new PR then.

**Comment by aaomidi** [comment]
Thank you both, I've opened a new PR :)

**Pull request closed without merging**